Support atomic_fence instrinstic

I'd love to try `miri` on [`evmap`](https://github.com/jonhoo/rust-evmap), but sadly I get the following error:

```
error[E0080]: Miri evaluation error: unimplemented intrinsic: atomic_fence
    --> /home/jon/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/sync/atomic.rs:2412:23
     |
2412 |             SeqCst => intrinsics::atomic_fence(),
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ Miri evaluation error: unimplemented intrinsic: atomic_fence
     |
note: inside call to `std::sync::atomic::fence` at src/write.rs:306:9
    --> src/write.rs:306:9
     |
306  |         atomic::fence(atomic::Ordering::SeqCst);
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Is this a known issue? Some cursory searching did not yield anything that seemed relevant.

--- a/src/shims/intrinsics.rs
+++ b/src/shims/intrinsics.rs
@@ -88,7 +88,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.write_scalar(val, place.into())?;
             }
 
-            "atomic_fence_acq" => {
+            "atomic_fence_acq" |
+            "atomic_fence_rel" |
+            "atomic_fence_acqrel" |
+            "atomic_fence" => {
                 // we are inherently singlethreaded and singlecored, this is a nop
             }
 

--- a/tests/run-pass/atomic.rs
+++ b/tests/run-pass/atomic.rs
@@ -1,9 +1,10 @@
-use std::sync::atomic::{AtomicBool, AtomicIsize, AtomicU64, Ordering::*};
+use std::sync::atomic::{fence, AtomicBool, AtomicIsize, AtomicU64, Ordering::*};
 
 fn main() {
     atomic_bool();
     atomic_isize();
     atomic_u64();
+    atomic_fences();
 }
 
 fn atomic_bool() {
@@ -57,6 +58,15 @@ fn atomic_u64() {
 
     ATOMIC.store(1, SeqCst);
     assert_eq!(ATOMIC.compare_exchange(0, 0x100, AcqRel, Acquire), Err(1));
-    assert_eq!(ATOMIC.compare_exchange_weak(1, 0x100, AcqRel, Acquire), Ok(1));
+    assert_eq!(
+        ATOMIC.compare_exchange_weak(1, 0x100, AcqRel, Acquire),
+        Ok(1)
+    );
     assert_eq!(ATOMIC.load(Relaxed), 0x100);
+}
+
+fn atomic_fences() {
+    fence(SeqCst);
+    fence(Release);
+    fence(Acquire);
 }

